doc/how_to/test/test.md

```diff
@@ -0,0 +1,405 @@
+    app.run_delay=0.3
+
+    def run():
+        app.run=True
+
+    benchmark(run)
+    assert benchmark.stats['min'] >= 0.3
+    assert benchmark.stats['max'] < 0.4
+```
+
+Lets run `pytest test_app.py`.
+
+```bash
+$ pytest test_app.py
+======================================================================= test session starts
+
+collected 4 items
+
+test_app.py ....                                                                                                                                             [100%]
+
+
+------------------------------------------------- benchmark: 1 tests ------------------------------------------------
+Name (time in ms)             Min       Max      Mean  StdDev    Median     IQR  Outliers     OPS  Rounds  Iterations
+---------------------------------------------------------------------------------------------------------------------
+test_run_performance     308.4448  311.8413  310.1718  1.2477  310.1379  1.5623       2;0  3.2240       5           1
+---------------------------------------------------------------------------------------------------------------------
+
+Legend:
+  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
+  OPS: Operations Per Second, computed as 1 / Mean
+================================================================== 4 passed in 4.82s
+```
+
+Notice how we used the `benchmark` fixture of [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/) to test the performance of the `run` event.
+
+## Test the frontend with pytest and Playwright
+
+Sometimes, for example when you create [custom components](https://panel.holoviz.org/user_guide/Custom_Components.html), it can be useful to test the frontend behaviour.
+
+For testing the frontend we recommend the framework [Playwright](https://playwright.dev/) by Microsoft. Panel it self is tested by this framework.
+
+### Create the conftest.py
+
+The `conftest.py` file contains pytest fixtures. it will
+
+- provide us with an available `port`
+- clean up the Panel server after each test.
+
+Create the file `conftest.py` and add the code below.
+
+```python
+# conftest.py
+"""Shared configuration and fixtures for testing Panel"""
+import panel as pn
+import pytest
+
+PORT = [6000]
+
+@pytest.fixture
+def port():
+    PORT[0] += 1
+    return PORT[0]
+
+@pytest.fixture(autouse=True)
+def server_cleanup():
+    """
+    Clean up server state after each test.
+    """
+    try:
+        yield
+    finally:
+        pn.state.kill_all_servers()
+        pn.state._indicators.clear()
+        pn.state._locations.clear()
+        pn.state._templates.clear()
+        pn.state._views.clear()
+        pn.state._loaded.clear()
+        pn.state.cache.clear()
+        pn.state._scheduled.clear()
+        if pn.state._thread_pool is not None:
+            pn.state._thread_pool.shutdown(wait=False)
+            pn.state._thread_pool = None
```

Comment:
We should probably add some public method to do this.

Reply:
Cleaning up the state isn't only useful for UI tests, right?

Reply:
Its probably useful for any kind of test where you persist some state. For example if you set the `cache`.